tests/testsuite/pub_priv.rs

```diff
@@ -479,3 +479,63 @@ fn allow_priv_in_custom_build() {
         )
         .run()
 }
+
+#[cargo_test(nightly, reason = "exported_private_dependencies lint is unstable")]
+fn recursive_priv_warning() {
+    Package::new("priv_dep", "0.1.0")
+        .file("src/lib.rs", "pub struct FromPriv;")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            cargo-features = ["public-dependency"]
+
+            [package]
+            name = "foo"
+            version = "0.0.1"
+
+            [dependencies]
+            priv_dep = "0.1.0"
+        "#,
+        )
+        .file(
+            "src/lib.rs",
+            "
+            mod grandparent_bar;
+            pub use grandparent_bar::*;
+        ",
+        )
+        .file(
+            "src/grandparent_bar/mod.rs",
+            "
+            mod parent_bar;
+            pub use self::parent_bar::*;
+        ",
+        )
+        .file(
+            "src/grandparent_bar/parent_bar/mod.rs",
+            "
+            mod pub_bar;
+            pub use self::pub_bar::*;
+        ",
+        )
+        .file(
+            "src/grandparent_bar/parent_bar/pub_bar/mod.rs",
+            "
+            extern crate priv_dep;
+            pub fn use_priv(_: priv_dep::FromPriv) {}
+        ",
+        )
```

Comment:
This seems to be showing how things work with children modules, rather than children packages.

We also need to be focusing on verifying that the recursively public API does not cause warnings.  One caution about it is that rustc does not correctly lint for re-exports.

Reply:
> One caution about it is that rustc does not correctly lint for re-exports.

`recursive_package_pub_no_warning` and `recursive_module_pub_no_warning`  had added newly,  I thinks this two case can be prove `rustc` handle the lint correctly.

Reply:
> One caution about it is that rustc does not correctly lint for re-exports.

`recursive_package_pub_no_warning` and `recursive_package_priv_warning` had added newly.

And if my testcase write correctly, the `recursive_package_pub_no_warning` warning by rustc is definitely wrong.

Reply:
`recursive_package_priv_warning` is similar to this issue: https://github.com/rust-lang/cargo/issues/13193